### ISSUE TITLE
[Q] How To Efficiently Log Moving Average

### ISSUE THREAD

**Issue by jhkonan** [wandb/wandb#4014]
I am new to wandb. I would like to simplify logging as much as possible, ideally limiting the need for too much logging code. The examples I create will be shared with my students, so the less lines to explain -- the better.

Here is what my current train loop looks like. 

```
def train(encoder, decoder, encoder_optim, decoder_optim, criterion, iterable):

    wandb.watch(encoder, criterion, log="all", log_freq=5000)
    wandb.watch(decoder, criterion, log="all", log_freq=5000)

    encoder.train()
    decoder.train()

    ### Initialize Log
    moving_total = 0
    moving_average = []

    for iter, (input_tensor, target_tensor) in enumerate(iterable):

        ### Forward Propagation & Loss Calculation
        loss = train_iter(input_tensor, target_tensor, encoder, decoder, criterion)

        ### Initialize Gradients
        encoder_optim.zero_grad()
        decoder_optim.zero_grad()

        ### Backward Propagation
        loss.backward()

        ### Gradient Descent
        encoder_optim.step()
        decoder_optim.step()

        ### Log Results
        moving_total += loss.item() / len(target_tensor)
        if (iter + 1) % 5000 == 0:
            print("{} | Loss {:.4f}".format(iter + 1, moving_total / 5000))
            wandb.log({"Loss": moving_total / 5000}, step=iter+1)
            moving_total = 0

    return moving_average
```

Is it possible to consolidate "Initialize Log" and "Log Results"? If log_freq is too small, then training gets really slow. Ideally it would function as implemented above, where a list is constructed until loq_freq is met. I want to use only in-built wandb accounting. It seems to defeat the purpose of the framework to do the accounting externally -- which makes me think I am using it wrong.

Thank you for your suggestions and feedback.

**Comment by ramit-wandb** [comment]
Hi @jhkonan!

We provide tools on the web platflorm itself to perform running averages which can be used to smooth your data. You can read more about them here : https://docs.wandb.ai/ref/app/features/panels/line-plot/smoothing#running-average

Using the online implementation we have for smoothing will let you simply remove the Initialize log section of your code.

**Comment by jhkonan** [comment]
The problem with your suggested approach is that, if we log every iteration, one loop will take 6x as long to execute.

The ideal solution averages after log_freq (or some other parameter), calculates the statistic, then logs the statistic to the server.

**Comment by ramit-wandb** [comment]
Hey @jhkonan,

In that case, I think how you have set up your code is the best way to go about this. You will definitely need some kind of data structure to store your moving averages over time, a list does the job well here.

**Comment by jhkonan** [comment]
Is there a way to separate the model logging from the loss logging?

I think logging the loss every iteration may be okay, but the model logging needs to be done less frequently.

**Comment by ramit-wandb** [comment]
Are you talking about the model gradients/parameters that are logged through `wandb.watch`? There does not seem to be any model logging code other than that. You should be able to reduce the number of times they are logged by increasing the `log_freq` parameter.

**Comment by jhkonan** [comment]
When only `wandb.log` Is used, I am able to get the expected loss plot in a reasonable amount of time. In this case, the suggestion of using smoothing within wandb works with little overhead.

If I include `wandb.watch` before the loop, then, every time `wandb.log` is called, the gradients are being uploaded and there is a massive overhead.

I want the gradients to be logged less often than the loss, which is not the case in the provided workflow.

**Comment by ramit-wandb** [comment]
Hey @jhkonan,

Got it. In case you want to reduce the number of times the gradient is uploaded, you can increase the value of `log_freq`.

Thanks,
Ramit

**Comment by ramit-wandb** [comment]
Hi @jhkonan, since we have not heard back from you we are going to close this request. If you would like to re-open the conversation, please let us know!
​
